src/hotspot/os/windows/memMapPrinter_windows.cpp

```diff
@@ -0,0 +1,248 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+
+#include "nmt/memMapPrinter.hpp"
+#include "runtime/os.hpp"
+
+#include <limits.h>
+#include <winnt.h>
+#include <memoryapi.h>
+#include <psapi.h>
+
+/* maximum number of mapping records returned */
+static const int MAX_REGIONS_RETURNED = 1000000;
+
+class MappingInfo {
+public:
+  stringStream _ap_buffer;
+  stringStream _state_buffer;
+  stringStream _protect_buffer;
+  stringStream _type_buffer;
+  char _file_name[MAX_PATH];
+
+  MappingInfo() {}
+
+  void process(MEMORY_BASIC_INFORMATION& mem_info) {
+    _ap_buffer.reset();
+    _state_buffer.reset();
+    _protect_buffer.reset();
+    _type_buffer.reset();
+    get_protect_string(_ap_buffer, mem_info.AllocationProtect);
+    get_state_string(_state_buffer, mem_info);
+    get_protect_string(_protect_buffer,  mem_info.Protect);
+    get_type_string(_type_buffer, mem_info);
+    _file_name[0] = 0;
+    if (mem_info.Type == MEM_IMAGE) {
+      HMODULE hModule = 0;
+      if (GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS, static_cast<LPCSTR>(mem_info.AllocationBase), &hModule)) {
+        GetModuleFileName(hModule, _file_name, sizeof(_file_name));
+      }
+    }
+  }
+
+  void get_protect_string(outputStream& out, DWORD prot) {
+    const char read_c = prot & (PAGE_READONLY | PAGE_READWRITE | PAGE_EXECUTE_READ | PAGE_EXECUTE_READWRITE | PAGE_WRITECOPY | PAGE_EXECUTE_WRITECOPY) ? 'r' : '-';
+    const char write_c = prot & (PAGE_READWRITE | PAGE_WRITECOPY | PAGE_EXECUTE_READWRITE | PAGE_EXECUTE_WRITECOPY) ? 'w' : '-';
+    const char execute_c = prot & (PAGE_EXECUTE | PAGE_EXECUTE_READ | PAGE_EXECUTE_READWRITE | PAGE_EXECUTE_WRITECOPY) ? 'x' : '-';
+    out.print("%c%c%c", read_c, write_c, execute_c);
+    if (prot & (PAGE_WRITECOPY | PAGE_EXECUTE_WRITECOPY)) {
+      out.put('c');
+    }
+    if (prot & PAGE_GUARD) {
+      out.put('g');
+    }
+    if (prot & PAGE_NOCACHE) {
+      out.put('n');
+    }
+    if (prot & PAGE_WRITECOMBINE) {
+      out.put('W');
+    }
+    const DWORD bits = PAGE_NOACCESS | PAGE_READONLY | PAGE_READWRITE | PAGE_EXECUTE_READ | PAGE_EXECUTE_READWRITE
+                        | PAGE_WRITECOPY | PAGE_EXECUTE_WRITECOPY | PAGE_EXECUTE
+                        | PAGE_GUARD | PAGE_NOCACHE | PAGE_WRITECOMBINE;
+    assert((prot & bits) == prot, "Unknown Windows memory protection value: 0x%x unknown bits: 0x%x", prot, prot & ~bits);
+  }
+
+  void get_state_string(outputStream& out, MEMORY_BASIC_INFORMATION& mem_info) {
+    if (mem_info.State == MEM_COMMIT) {
+      out.put('c');
+    } else if (mem_info.State == MEM_FREE) {
+      out.put('f');
+    } else if (mem_info.State == MEM_RESERVE) {
+      out.put('r');
+    } else {
+      fatal("Unknown Windows memory state value: 0x%x", mem_info.State);
```

Comment:
I was thinking about this - my original intent was to be fatal only on a debug build, not release.  I have added a message in the output, and changed these to warnings (for both release and debug)

Reply:
I would prefer an assert here. It will crash in debug builds, which is fine, since if it crashes it is either a bug or an hitherto unknown state. In any case something that should be fixed.

Reply:
I believe we can all agree that fatal() was too harsh.
Is there a macro that is the equivalent of "assert(false, ....)"?  
my current unchecked in code has 'warning()' which will log the problem.  I am happy to replace this with assert().
How do you feel about a message in the output even in release builds?